dask/dataframe/groupby.py

```diff
@@ -279,37 +301,117 @@ def _var_combine(g, levels):
 def _var_agg(g, levels, ddof):
     g = g.groupby(level=levels, sort=False).sum()
     nc = len(g.columns)
-    x = g[g.columns[:nc // 3]]
-    x2 = g[g.columns[nc // 3:2 * nc // 3]].rename(columns=lambda c: c[:-3])
-    n = g[g.columns[-nc // 3:]].rename(columns=lambda c: c[:-6])
+    x = g[g.columns[: nc // 3]]
+    x2 = g[g.columns[nc // 3 : 2 * nc // 3]].rename(columns=lambda c: c[:-3])
+    n = g[g.columns[-nc // 3 :]].rename(columns=lambda c: c[:-6])
 
     # TODO: replace with _finalize_var?
     result = x2 - x ** 2 / n
-    div = (n - ddof)
+    div = n - ddof
     div[div < 0] = 0
     result /= div
     result[(n - ddof) == 0] = np.nan
     assert is_dataframe_like(result)
     return result
 
 
+def _cov_combine(g, levels):
+    return g
+
+
+def _cov_finalizer(df, cols):
+    vals = []
+    num_elements = len(list(it.product(cols, repeat=2)))
+    num_cols = len(cols)
+    vals = np.zeros(num_elements)
+    col_idx_mapping = dict(zip(cols, range(len(cols))))
+    for i, j in it.combinations_with_replacement(df[cols].columns, 2):
+        x = col_idx_mapping[i]
+        y = col_idx_mapping[j]
+        idx = x + num_cols * y
+        mul_col = "%s%s" % (i, j)
+        ni = df["%s-count" % i]
+        nj = df["%s-count" % j]
+
+        n = np.sqrt(ni * nj)
+        div = n - 1
+        div[div < 0] = 0
+
+        val = (df[mul_col] - df[i] * df[j] / n).values[0] / div
+        vals[idx] = val
+        if i != j:
+            idx = num_cols * x + y
+            vals[idx] = val
+    level_1 = cols
+    index = pd.MultiIndex.from_product([level_1, level_1])
+    return pd.Series(vals, index=index)
+
+
+def mul_cols(df, cols):
+    _df = pd.DataFrame()
+    for i, j in it.combinations_with_replacement(df[cols].columns, 2):
+        col = "%s%s" % (i, j)
+        _df[col] = df[i] * df[j]
+    return _df
+
+
+def _cov_chunk(df, *index):
+    if is_series_like(df):
+        df = df.to_frame()
+
+    df = df.copy()
+    cols = df._get_numeric_data().columns
+    cols = cols.drop(*index)
+
+    g = _groupby_raise_unaligned(df, by=index)
+    x = g.sum()
+    mul = g.apply(mul_cols, cols=cols).reset_index(level=1, drop=True)
+    n = g[x.columns].count().rename(columns=lambda c: c + "-count")
+
+    return (x, mul, n)
+
+
+def _cov_agg(t, levels, ddof):
+    xs = []
+    muls = []
+    ns = []
+    cols = t[0][0].columns
+    for x, mul, n in t:
+        xs.append(x)
+        muls.append(mul)
+        ns.append(n)
+
+    x_sums = pd.concat(xs).groupby(level=levels, sort=False).sum()
+    x_muls = pd.concat(muls).groupby(level=levels, sort=False).sum()
+    x_ns = pd.concat(ns).groupby(level=levels).sum()
```

Comment:
At some point we might want to use our dispatched concat function instead, `dask.dataframe.methods.concat_dispatched`.  Not particularly important yet though.